src/lib/flow/sol-flow-js.c

```diff
@@ -0,0 +1,1422 @@
+flow_js_get_port_out(const struct sol_flow_node_type *type, uint16_t port)
+{
+    const struct flow_js_type *js_type = (struct flow_js_type *)type;
+    const struct flow_js_port_out *p = sol_vector_get(&js_type->ports_out, port);
+
+    return p ? &p->type : NULL;
+}
+
+#ifdef SOL_FLOW_NODE_TYPE_DESCRIPTION_ENABLED
+static const struct sol_flow_node_type_description sol_flow_node_type_js_description = {
+    .api_version = SOL_FLOW_NODE_TYPE_DESCRIPTION_API_VERSION,
+    .name = "js",
+    .category = "js",
+    .symbol = "SOL_FLOW_NODE_TYPE_JS",
+    .options_symbol = "sol_flow_node_type_js_options",
+    .version = NULL,
+    /* TODO: Add a way for the user specify description, author, url and license. */
+};
+
+static int
+setup_description(struct flow_js_type *type)
+{
+    struct sol_flow_node_type_description *desc;
+    struct sol_flow_port_description **p;
+    struct flow_js_port_in *port_type_in;
+    struct flow_js_port_out *port_type_out;
+    int i, j;
+
+    type->base.description = calloc(1, sizeof(struct sol_flow_node_type_description));
+    SOL_NULL_CHECK(type->base.description, -ENOMEM);
+
+    type->base.description = memcpy((struct sol_flow_node_type_description *)type->base.description,
+        &sol_flow_node_type_js_description, sizeof(struct sol_flow_node_type_description));
+
+    desc = (struct sol_flow_node_type_description *)type->base.description;
+
+    /* Extra slot for NULL sentinel at the end. */
+    desc->ports_in = calloc(type->ports_in.len + 1, sizeof(struct sol_flow_port_description *));
+    SOL_NULL_CHECK_GOTO(desc->ports_in, fail_ports_in);
+
+    p = (struct sol_flow_port_description **)desc->ports_in;
+    for (i = 0; i < type->ports_in.len; i++) {
+        p[i] = calloc(1, sizeof(struct sol_flow_port_description));
+        SOL_NULL_CHECK_GOTO(p[i], fail_ports_in_desc);
+
+        port_type_in = sol_vector_get(&type->ports_in, i);
+
+        p[i]->name = port_type_in->name;
+        p[i]->description = "Input port";
+        p[i]->data_type = port_type_in->type_name;
+        p[i]->array_size = 0;
+        p[i]->base_port_idx = i;
+        p[i]->required = false;
+    }
+
+    /* Extra slot for NULL sentinel at the end. */
+    desc->ports_out = calloc(type->ports_out.len + 1, sizeof(struct sol_flow_port_description *));
+    SOL_NULL_CHECK_GOTO(desc->ports_out, fail_ports_in_desc);
+
+    p = (struct sol_flow_port_description **)desc->ports_out;
+    for (j = 0; j < type->ports_out.len; j++) {
+        p[j] = calloc(1, sizeof(struct sol_flow_port_description));
+        SOL_NULL_CHECK_GOTO(p[j], fail_ports_out_desc);
+
+        port_type_out = sol_vector_get(&type->ports_out, j);
+
+        p[j]->name = port_type_out->name;
+        p[j]->description = "Output port";
+        p[j]->data_type = port_type_out->type_name;
+        p[j]->array_size = 0;
+        p[j]->base_port_idx = j;
+        p[j]->required = false;
+    }
+
+    return 0;
+
+fail_ports_out_desc:
+    if (j > 0) {
```

Comment:
Since these loops can `goto` fail_ports_in/out_desc at any iteration, I keep the last iteration, in this case `i` and `j`, to iterate through this arrays freeing the values. But yes, I can remove the `if`.